provider/ovh.go

```diff
@@ -0,0 +1,354 @@
+}
+
+// Records returns the list of records in all relevant zones.
+func (p *OVHProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, error) {
+
+	_, records, err := p.zonesRecords(ctx)
+	if err != nil {
+		return nil, err
+	}
+	endpoints := ovhGroupByNameAndType(records)
+	log.Infof("OVH: %d endpoints have been found\n", len(endpoints))
+	return endpoints, nil
+}
+
+// ApplyChanges applies a given set of changes in a given zone.
+func (p *OVHProvider) ApplyChanges(ctx context.Context, changes *plan.Changes) error {
+	zones, records, err := p.zonesRecords(ctx)
+	zonesChangeUniques := map[string]bool{}
+	if err != nil {
+		return err
+	}
+
+	allChanges := make([]ovhChange, 0, countTargets(changes.Create, changes.UpdateNew, changes.UpdateOld, changes.Delete))
+
+	allChanges = append(allChanges, newOvhChange(ovhCreate, changes.Create, zones, records)...)
+	allChanges = append(allChanges, newOvhChange(ovhCreate, changes.UpdateNew, zones, records)...)
+
+	allChanges = append(allChanges, newOvhChange(ovhDelete, changes.UpdateOld, zones, records)...)
+	allChanges = append(allChanges, newOvhChange(ovhDelete, changes.Delete, zones, records)...)
+
+	log.Infof("OVH: %d changes will be done\n", len(allChanges))
+
+	eg, _ := errgroup.WithContext(ctx)
+	for _, change := range allChanges {
+		change := change
+		zonesChangeUniques[change.Zone] = true
+		eg.Go(func() error { return p.change(change) })
+	}
+	if err := eg.Wait(); err != nil {
+		return err
+	}
+
+	log.Infof("OVH: %d zones will be refreshed\n", len(zonesChangeUniques))
+
+	eg, _ = errgroup.WithContext(ctx)
+	for zone := range zonesChangeUniques {
+		zone := zone
+		eg.Go(func() error { return p.refresh(zone) })
+	}
+	if err := eg.Wait(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *OVHProvider) refresh(zone string) error {
+	log.Debugf("OVH: Refresh %s zone\n", zone)
+	return p.client.Post(fmt.Sprintf("/domain/zone/%s/refresh", zone), nil, nil)
+}
+
+func (p *OVHProvider) change(change ovhChange) error {
+	var err error = nil
+	switch change.Action {
+	case ovhCreate:
+		log.Debugf("OVH: Add an entry to %s\n", change.String())
+		err = p.client.Post(fmt.Sprintf("/domain/zone/%s/record", change.Zone), change.ovhRecordFields, nil)
+	case ovhDelete:
+		if change.ID == 0 {
+			err = ErrRecordToMutateNotFound
+			break
+		}
+		log.Debugf("OVH: Delete an entry to %s\n", change.String())
+		err = p.client.Delete(fmt.Sprintf("/domain/zone/%s/record/%d", change.Zone, change.ID), nil)
+	}
+	return err
+}
+
+func (p *OVHProvider) zonesRecords(ctx context.Context) ([]string, []ovhRecord, error) {
+	var allRecords []ovhRecord
+	zones, err := p.zones()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	chRecords := make(chan []ovhRecord, len(zones))
+	eg, ctx := errgroup.WithContext(ctx)
+	for _, zone := range zones {
+		zone := zone
+		eg.Go(func() error { return p.records(&ctx, &zone, chRecords) })
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, nil, err
+	}
+	close(chRecords)
```

Comment:
use defer close,  see https://gobyexample.com/defer

Reply:
Maybe i miss something, but this close a buffered channel, it while not deadlock if i don't close them before reading it ?